### PR TITLE
slice(.data, ...)

### PR DESCRIPTION
**Pull request by romainfrancois**
```r
> iris %>% group_by(Species) %>% slice(1, n())
# A tibble: 6 x 5
# Groups:   Species [3]
  Sepal.Length Sepal.Width Petal.Length Petal.Width Species   
         <dbl>       <dbl>        <dbl>       <dbl> <fct>     
1          5.1         3.5          1.4         0.2 setosa    
2          5           3.3          1.4         0.2 setosa    
3          7           3.2          4.7         1.4 versicolor
4          5.7         2.8          4.1         1.3 versicolor
5          6.3         3.3          6           2.5 virginica 
6          5.9         3            5.1         1.8 virginica 
```

### CODE REVIEW

File: R/tbl-df.r [hunk 1]
```diff
@@ -76,7 +76,16 @@ filter_.tbl_df <- function(.data, ..., .dots = list(), .preserve = TRUE) {
 
 #' @export
 slice.tbl_df <- function(.data, ...) {
-  slice_impl(.data, quos(...)[[1L]])
+  dots <- quos(...)
+  if (any(have_name(dots))) {
+    bad <- dots[have_name(dots)]
+    bad_eq_ops(bad, "must not be named, do you need `==`?")
```
Comment (krlmlr) [inline R/tbl-df.r]: How likely is `==` to appear legitimately in `slice()` ?

File: R/tbl-df.r [hunk 2]
```diff
@@ -76,7 +76,16 @@ filter_.tbl_df <- function(.data, ..., .dots = list(), .preserve = TRUE) {
...
+  } else if (is_empty(dots)) {
+    return(.data)
```
Comment (krlmlr) [inline R/tbl-df.r]: I wonder if `slice(.)` should be the same as `slice(integer())`, even if we break compatibility.

**Comment by romainfrancois** [comment]
Merging this now so that we have the right interface. I'll worry later about passing individual quosures down when I have clearer ideas about what hybrid evaluation can do for `slice` and `filter`

File: R/tbl-df.r (hunk 1 above)
Comment (romainfrancois) [reply R/tbl-df.r]: yeah, copy and paste thing

File: R/tbl-df.r (hunk 2 above)
Comment (romainfrancois) [reply R/tbl-df.r]: That makes sense. @hadley ?

Comment (romainfrancois) [reply R/tbl-df.r]: Something like this: 

```r
slice.tbl_df <- function(.data, ...) {
  dots <- quos(...)
  quosure <- if (is_empty(dots)) {
    quo(integer())
  } else {
    quo(c(!!!dots))
  }

  slice_impl(.data, quosure)
}
```

Comment (hadley) [reply R/tbl-df.r]: I think that makes it less compatible with base where `x[]` means select everything.

Comment (krlmlr) [reply R/tbl-df.r]: To me, the most important use cases for `x[]` are:

1. Assignment without changing attributes, e.g., `df[] <- lapply(df, ...)`
2. Subsetting only along one of two or more dimensions, e.g. `df[rows, ]`

Both don't seem relevant for `slice()` . On the other hand, adding a corner case for `slice(.)` might be bad for programming.

Comment (hadley) [reply R/tbl-df.r]: Nevertheless, the convention in base R is that empty subsetting yields the whole input. I don't think `slice()` is special enough to break that convention.

**Comment by romainfrancois** [comment]
@lionel- I'm struggling to merge this, and getting errors I'm not able to reproduce locally, e.g. https://travis-ci.org/tidyverse/dplyr/jobs/441733426

I'm wondering if this is because I'm doing something wrong with `slice.tbl_df` here: 

```r
slice.tbl_df <- function(.data, ...) {
  dots <- quos(...)
  if (any(have_name(dots))) {
    bad <- dots[have_name(dots)]
    bad_eq_ops(bad, "must not be named")
  } else if (is_empty(dots)) {
    return(.data)
  }

  quo <- quo(c(!!!dots))
  slice_impl(.data, quo)
}
```

especially, is `quo(c(!!!dots))` ok ?

**Comment by romainfrancois** [comment]
Ok, was "just" a tibble thing and had nothing to do with `slice()`. Should be fixed now.  

```
> test_check("dplyr")
── 1. Error: slice handles raw matrices (@test-slice.r#121)  ───────────────────
Column `b` must be a 1d atomic vector or a list
1: tibble(a = 1:4, b = matrix(as.raw(1:8), ncol = 2)) at testthat/test-slice.r:121
2: tibble::tibble(...) at /home/travis/build/tidyverse/dplyr/dplyr.Rcheck/tests/testthat/helper-dplyr.R:6
3: as_tibble(lst_quos(xs, expand = TRUE))
4: as_tibble.list(lst_quos(xs, expand = TRUE))
5: list_to_tibble(x, validate)
6: check_tibble(x)
7: invalid_df("must be [a] 1d atomic vector(s) or [a] list(s)", x, !is_1d)
8: stopc(pluralise_msg("Column(s) ", vars), " ", pluralise(problem, vars))
9: abort(paste0(...))
══ testthat results  ═══════════════════════════════════════════════════════════
OK: 3093 SKIPPED: 9 FAILED: 1
1. Error: slice handles raw matrices (@test-slice.r#121) 
Error: testthat unit tests failed
In addition: Warning message:
call dbDisconnect() when finished working with a connection 
Execution halted
```

**Pull request merged**